README.md
# MMM-KVV
<B>Station monitor</B> for the <B>Karlsruhe local transport (KVV)</B> bus, tram and train system.<P>

This module is an extension of the amazing [MagicMirror<sup>2</sup>](https://github.com/MichMich/MagicMirror) project by [MichMich](https://github.com/MichMich/) which has inspired me to share my coding skills with others as well. Check it out, you know you want to :). <P>

It's always nice to see mirrors using my work, so feel free to send me some screenshots of your implementations.<P>

Lastly, why not join in on our discussions at the official [MagicMirror<sup>2</sup> Forum](http://forum.magicmirror.builders/)?

## Screenshots

![German version](screenshots/screenshot_de.png)
![English version](screenshots/screenshot_en.png)

## Current version

v1.0.3

## Languages
As of version 1.0.0, MMM-KVV features language support for `German (de)` and `English (en)` mirrors.

## Prerequisite
A working installation of [MagicMirror<sup>2</sup>](https://github.com/MichMich/MagicMirror)
 
## Dependencies
  * npm
  * [request](https://www.npmjs.com/package/request)

## Installation
1. Navigate into your MagicMirror's `modules` folder.
2. Execute `git clone https://github.com/yo-less/MMM-KVV.git`.
3. Execute `cd MMM-KVV`.
3. Execute `npm install`.

## Module behavior
Please note that this module auto-creates a module header which displays the name of the chosen Karlsruhe local transport stop. It is therefore recommended not to add a 'header' entry to your config.js for this module.<P>
This module automatically disappears from your mirror as soon as a station has stopped offering connections at night. It reappears as soon as your chosen station is scheduled to be served again.<P>
This module has been programmed to allow for multiple instances. Simply add more MMM-KVV config entries to your config.js file to display multiple stations and configure them according to your needs.

## Configuration
Sample minimum configuration entry for your `~/MagicMirror/config/config.js`:

    ...
    
    {
        module: 'MMM-KVV',
        position: 'top_left',
        config: {
			stopID: '',		// Which stop would you like to have displayed?			
        }
    } 						// If this isn't your last module, add a comma after the bracket
    
    ...

Sample configuration entry for your `~/MagicMirror/config/config.js` with optional parameters:

    ...
    
    {
        module: 'MMM-KVV',
        position: 'top_left',
        config: {
			stopID: '',		// Which stop would you like to have displayed?
			maxConn: 6,		// How many connections would you like to see? (Maximum: 10)
			lines: '',		// What lines are you interested in?
			direction: '',	// If you want to, you can limit the information to one of two directions
			labelRow: true, // Show or hide column headers
		    reload: 60000 	// How often should the information be updated? (In milliseconds)
        }
    } 						// If this isn't your last module, add a comma after the bracket
    
    ...

## Figuring out the correct stopID
1. Open your web browser and navigate to the [KVV Live Web Page](http://live.kvv.de).
2. Use the search field to find the stop you are interested in.
3. Once you can see the list of connections for your stop in your browser, note the information after "stopId=",<BR> this is the `StopID` you are looking for.

## Config Options
| **Option** | **Default** | **Description** |
| :---: | :---: | --- |
| stopID | de:8212:89 | <BR>Which stop would you like to have displayed? <BR><EM> Default: Karlsruhe central station (tram stop)</EM><P> |
| maxConn<BR>`optional` | 8 | <BR> How many connections would you like to see? <BR><EM><B>Note</B>: The KVV API limits the maximum number of connections to 10.</EM><P> |
| lines<BR>`optional` |  | <BR> Only show connections for specific lines - use commas to choose multiple lines.<BR><EM> Example values: 'S1, 3'<BR><B>Note</B>: You <B>can</B> use spaces when setting this parameter in order to enhance legibility.</EM><P> |
| direction<BR>`optional` |  | <BR> There are really "two" stops to every stop, depending on what side of the street you're standing on. You can limit the presented information to one of those "two" stops.<BR> <EM>Possible values: 1, 2</EM><P> |
| labelRow<BR>`optional` | true | <BR> Show or hide column headers<BR> <EM>Possible values: true, false</EM><P> |
| reload<BR>`optional`  | 60000 | <BR> How often should the information be updated? (In milliseconds) <BR><EM> Default: Every minute </EM><P> |

## Licence
MIT License

Copyright (c) 2016 yo-less (https://github.com/yo-less/)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


changelog.md
# MMM-KVV changelog
This module adheres to [Semantic Versioning](http://semver.org/).

## [1.0.3] - 2016-10-21

### Fixed
- Restored connection by switching API url prefix to "https://"

### Added
- Version info now shown inside 'MMM-KVV.js'

## [1.0.2] - 2016-09-03

### Changed
- "Departure" changed to "Due" in English translation file (en.json)
- Due time is given as time phrase: '2 min' -> 'In 2 minutes'
- Increased space between 'line' and 'destination' columns
- Node helper now reports back with the console line "Starting node helper for: MMM-KVV"

## [1.0.1] - 2016-08-29

### Added
- Animation upon module disappearance
- Animation upon module reappearance
- Short text advertizing no connections shortly before module disappears

### Fixed
- Module failed to reappear once it had disappeared

## [1.0.0] - 2016-08-25 
### Initial release

node_helper.js
/* Magic Mirror
 * Module: MMM-KVV
 *
 * By yo-less
 * MIT Licensed.
 */

const request = require('request');
const NodeHelper = require("node_helper");



module.exports = NodeHelper.create({

    start: function() {
        console.log("Starting node helper for: " + this.name);
    },

	
	/* getParams
	 * Generates an url with api parameters based on the config.
	 *
	 * return String - URL params.
	 */
	
	getParams: function() {
			var params = this.config.stopID;
			params += "?";
			if (this.config.maxConn !== '') {
			params += "maxInfos=" + this.config.maxConn + "&";
			}
			params +="key=" + this.config.apiKey;
			return params;
	},
	
    socketNotificationReceived: function(notification, payload) {
        if(notification === 'CONFIG'){
            this.config = payload;
			var kvv_url = this.config.apiBase + this.getParams();
			this.getData(kvv_url, this.config.stopID);
        }
    },

    getData: function(options, stopID) {
		request(options, (error, response, body) => {
	        if (response.statusCode === 200) {
				this.sendSocketNotification("TRAMS" + stopID, JSON.parse(body));
				} else {
                console.log("Error getting tram connections " + response.statusCode);
            }
        });
    }
});

MMM-KVV.js
/* Magic Mirror
 * Module: MMM-KVV
 *
 * By yo-less / https://github.com/yo-less
 * MIT Licensed.
 * 
 * v1.0.3
 */

Module.register("MMM-KVV", {

    defaults: {
		apiBase: 'https://live.kvv.de/webapp/departures/bystop/',
		apiKey: '377d840e54b59adbe53608ba1aad70e8',
		stopID: 'de:8212:89',
		maxConn: '8',
		lines: '',
		direction: '',
		labelRow: true,
		stopName: 'KVV',
        reload: 1 * 60 * 1000       // every minute
    },

    getTranslations: function () {
        return {
            en: "translations/en.json",
            de: "translations/de.json"
        };
    },

    getStyles: function () {
        return ["MMM-KVV.css"];
    },

    start: function () {
		var self = this;
        Log.info("Starting module: " + this.name);
        this.sendSocketNotification("CONFIG", this.config);
		setInterval(
			function()
			{self.sendSocketNotification("CONFIG", self.config);}
			,this.config.reload);
    },

		
    socketNotificationReceived: function (notification, payload) {
		if (notification === "TRAMS" + this.config.stopID) {
			this.kvv_data = payload;
			this.config.stopName = payload.stopName;
			this.updateDom();			
	    }
	},

    getDom: function () {
					
		// Auto-create MagicMirror header

		var wrapper = document.createElement("div");
        var header = document.createElement("header");
        header.innerHTML = this.config.stopName;
        wrapper.appendChild(header);
	
		// Loading data notification
		
	    if (!this.kvv_data) {
            var text = document.createElement("div");
            text.innerHTML = this.translate("LOADING");
            text.className = "small dimmed";
            wrapper.appendChild(text);
        
		} else {
			
			// Start creating connections table
			
			var table = document.createElement("table");
			table.classList.add("small", "table");
			table.border='0';
				
							
			// Check if there are any connections before adding a label row
				
				var counter = 0;
				
				for (var f in this.kvv_data.departures){
					
					var tram = this.kvv_data.departures[f];
					
					if(this.config.lines !== '' ) {
					
						if(this.kvv_lines(tram.route, this.config.lines)) {
																		
									if (this.config.direction === '1') {
										if (tram.direction === '1') {
											counter = counter + 1;
										}
									} else if (this.config.direction === '2') {
										if (tram.direction === '2') {
											counter = counter + 1;
										}
									} else {
										counter = counter + 1;
									}
						} 
					
					} else {
						
						if (this.config.direction === '1') {
								if (tram.direction === '1') {
										counter = counter + 1;
								}
							} else if (this.config.direction === '2') {
								if (tram.direction === '2') {
										counter = counter + 1;
								}
							} else {
								counter = counter + 1;
							}
					}
					
					}
									
				if (counter > 0 && this.config.labelRow) {
					table.appendChild(this.createLabelRow());
				}
				
				table.appendChild(this.createSpacerRow());
					
			// Listing selected connections
				var counter = 0;
				
				for (var f in this.kvv_data.departures){
					
					var tram = this.kvv_data.departures[f];
					
					if(this.config.lines !== '' ) {
					
						if(this.kvv_lines(tram.route, this.config.lines)) {
																		
									if (this.config.direction === '1') {
										if (tram.direction === '1') {
												table.appendChild(this.createDataRow(tram));
												counter = counter + 1;
										}
									} else if (this.config.direction === '2') {
										if (tram.direction === '2') {
												table.appendChild(this.createDataRow(tram));
												counter = counter + 1;
										}
									} else {
										table.appendChild(this.createDataRow(tram));
										counter = counter + 1;
									}
						} 
					
					} else {
						
						if (this.config.direction === '1') {
								if (tram.direction === '1') {
										table.appendChild(this.createDataRow(tram));
										counter = counter + 1;
								}
							} else if (this.config.direction === '2') {
								if (tram.direction === '2') {
										table.appendChild(this.createDataRow(tram));
										counter = counter + 1;
								}
							} else {
								table.appendChild(this.createDataRow(tram));
								counter = counter + 1;
							}
					}
					
					}
									
				if (counter == 0) {
									
					if (!this.hidden) {
						table.appendChild(this.createNoTramRow());
						wrapper.appendChild(table);
						this.hide(10000);
										}
					
				} else {

					if (this.hidden) {
						this.show(5000);
					} 
				
					wrapper.appendChild(table);
				}
					
			}
					
		return wrapper; 
		
	},
	
	// Function checking whether a line is wanted by the user
	
	kvv_lines: function(queried_line, lines_config) {
	
		//Remove spaces in the config parameter (if any)
		var lines_clean = lines_config.replace(/\s+/g,'');
		
		//Create a line array from the config parameter
		var lines_array = lines_clean.split(",");
		
		//Check if the line in question is in the config
		
		for (var i=0;i<lines_array.length;i++) {
			if(lines_array[i] == queried_line)
			{return true;}
		}
	return false;
	},

    createLabelRow: function () {
        var labelRow = document.createElement("tr");
		
        var lineLabel = document.createElement("th");
		lineLabel.className = "line";
        lineLabel.innerHTML = this.translate("LINE");
        labelRow.appendChild(lineLabel);

        var destinationLabel = document.createElement("th");
		destinationLabel.className = "destination";
        destinationLabel.innerHTML = this.translate("DESTINATION");
        labelRow.appendChild(destinationLabel);

        var departureLabel = document.createElement("th");
		departureLabel.className = "departure";
        departureLabel.innerHTML = this.translate("DEPARTURE");
        labelRow.appendChild(departureLabel);
		
		return labelRow;
    },

    createSpacerRow: function () {
        var spacerRow = document.createElement("tr");
		
		var spacerHeader = document.createElement("th");
		spacerHeader.className = "spacerRow";
		spacerHeader.setAttribute("colSpan", "3");
		spacerHeader.innerHTML = "";
		spacerRow.appendChild(spacerHeader); 
      	
		return spacerRow;
    },

	
    createNoTramRow: function () {
        var noTramRow = document.createElement("tr");
		
		var noTramHeader = document.createElement("th");
		noTramHeader.className = "noTramRow";
		noTramHeader.setAttribute("colSpan", "3");
		noTramHeader.innerHTML = this.translate("NO-TRAMS");
		noTramRow.appendChild(noTramHeader); 
      	
		return noTramRow;
    },
	
    createDataRow: function (data) {
        var row = document.createElement("tr");

        var line = document.createElement("td");
		line.className = "line";
        line.innerHTML = data.route;
        row.appendChild(line);

        var destination = document.createElement("td");
        destination.innerHTML = data.destination;
        row.appendChild(destination);

        var departure = document.createElement("td");
		departure.className = "departure";
        if (data.time == "0"){
		departure.innerHTML = this.translate("NOW");
		} else if (data.time.charAt(2)==":"){	// Append "Uhr" to data given in "hh:mm" format if config language is German
		departure.innerHTML = data.time + this.translate("TIME");
		} else if (data.time.substr(data.time.length - 5) == "1 min") {
		departure.innerHTML = 'In ' + data.time.slice(0, -4) + ' ' + this.translate("MINUTE"); // Give remaining time as 'In 1 minute' rather than as '1 min'
		} else if (data.time.substr(data.time.length - 3) == "min") {
		departure.innerHTML = 'In ' + data.time.slice(0, -4) + ' ' + this.translate("MINUTES"); // Give remaining time as 'in d minutes' rather than as '\d min'
		} else {
		departure.innerHTML = data.time;
		}
        row.appendChild(departure);

        return row;
    }

});